### PR TITLE
feat: Implemented 1 validator consensus for the main node

### PR DESCRIPTION
**Pull request by pompon0**
## What ❔

Implemented 1 validator consensus for the main node.
It is NOT enabled yet, corresponding configuration still needs to be provided.
It is responsible for populating consensus column in the Miniblocks table.
Consensus will be running asynchronously for now, not affecting the miniblock production at all.

Fixes BFT-388.

## Why ❔

Reaching a consensus with just 1 validator doesn't provide much value, however we want to start running the consensus code in production (in a "shadow" mode) to make sure it works as expected before we start relying on it.

### CODE REVIEW

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/storage/mod.rs:140]: Just to understand better: Is there a risk here that a non-transactional update would result in a logical error, or is it just a best practice of wrapping related DB calls in a transaction?

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/storage/mod.rs:63]: Nit: `'a` can be elided here and in `fetch_block` (i.e., you can write `txn: StorageProcessor<'_>`).

File: core/lib/zksync_core/src/consensus/wrap_error.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/wrap_error.rs:10]: Nit: This implementation is used in 2 places (`impl WrapError for StorageError` and a single place in `put_block`), and can be replaced with `.context(_)`. IMO, it could make sense to specialize this extension for `StorageError` / `StorageResult` specifically and maybe rename methods to something like `db_context()` / `with_db_context()`.

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage/mod.rs:254]: yes, this is for single-validator only. multi-validator will require interaction with StateKeeper and will look totally different.

File: core/lib/zksync_core/src/consensus/wrap_error.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/wrap_error.rs:10]: I'd rather prefer to generalize it, so that WrapError can be implemented for an arbitrary error type which accumulates context in a nontrivial way. Also StorageError ~ Canceled + Internal, a type which IMO is as applicable as anyhow::Error, so I think it would be useful to move StorageError to concurrency crate and rename it.

Comment (pompon0) [reply core/lib/zksync_core/src/consensus/wrap_error.rs:10]: anyhow::Error needs some helper to support WrapError, because (as mentioned above) it doesn't implement anyhow::Context directly. This helper doesn't necessarily have to be WrapError implementation, but it would look quite consistent IMO.

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage/mod.rs:126]: yes, I suppose so. Given that the high level logic is quite different for these modules I haven't paid any attention to deduplicate helper functions so far.

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage/mod.rs:209]: find_head_number takes time linear to the number of miniblocks without consensus field set, which can get quite large, especially initially when we start backfilling starting from a fixed genesis block. Note that we have a polling loop with this query, so it can get quite expensive. Sequential lookups on the other side will have amortized cost.

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage/mod.rs:140]: it is a read-write transaction and I don't want to rely on external assumptions about the storage access patterns if I don't have to.

File: core/lib/zksync_core/src/consensus/storage/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage/mod.rs:54]: done

File: core/lib/zksync_core/src/consensus/wrap_error.rs
Comment (slowli) [reply core/lib/zksync_core/src/consensus/wrap_error.rs:14]: It has inherent [`context()` method](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#method.context).

Comment (pompon0) [reply core/lib/zksync_core/src/consensus/wrap_error.rs:14]: thx, somehow I missed it

File: core/lib/zksync_core/src/consensus/wrap_error.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/wrap_error.rs]: my proposal: https://github.com/matter-labs/era-consensus/pull/43

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/storage.rs:262]: Nit: It could make sense to include `payload` / wanted payload into the error message for debugging purposes, although it would make to parse `payload` instead of encoding `want` in this case.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/storage.rs:340]: Note that I've fixed this place for the EN block store. Although in practice we won't call `missing_block_numbers` with args that would lead to a non-empty response, this is non-local reasoning, so I've found it too shaky.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/storage.rs:373]: Dumb question: Does the consensus code expect that this method can take an extended period of time to complete? I.e., could this block other logic from executing / completing?

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/testonly.rs:221]: Nit: IMO, `run_mock_metadata_calculator` would be a better name; otherwise, one could get an idea that a real `MetadataCalculator` should be run here.

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/testonly.rs:286]: Nit: You can use `send_replace` instead of `send` to signal that you allow for no receivers to be present.

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/tests.rs:26]: Bikeshedding: IMO, `const`s in the middle of the function body look weird; it's a partial case of [a Clippy lint](https://rust-lang.github.io/rust-clippy/master/index.html#/items_after_statements) that we don't have enabled.

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/tests.rs:42]: Dumb question: Can't errors be propagated here?

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/tests.rs:1]: Good job with testing 👍

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/testonly.rs:21]: Nit: You can probably reuse `MockMainNodeClient` from `sync_layer::tests` here; w/o generating blocks for it, it effectively works in the same way.

**Review by slowli — COMMENTED** [review]
I think there's an opportunity to share some main / test code with the EN-related consensus logic. I can do this in a follow-up PR after this one is merged.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:117]: yeah, that's from debugging, sry

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:154]: I'm not a big fan of macros and this explicit error-handling pattern feels ok to me. However if it starts to annoy  people at scale, we can implement that. In fact it should be possible to generalize anyhow::bail to any error type which implements `From<anyhow::Error>`.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:143]: I was thinking about that and I haven't reached a strong conclusion, I guess that we may want to stay on the safe side for now and forbid overwriting it altogether.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:277]: sure, I can add it

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:300]: I decided to keep the semantics simple and delegate "head updates" solely to run_background_tasks. This will prevent not noticing a hard-to-debug situation in which one forgets to call run_background_tasks and all methods keep working efficiently, except for subscribe() which works usually-unless-we-stop-calling-other-methods.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:373]: Good point. propose() may be blocking in case there is nothing to propose for a while(no txs in mempool for example). Still it shouldn't just give up proposing right away, but rather wait, timing out only if the consensus moves to another view already. With the current implementation it will block both replica and the leader actors from processing messages. Replica definitely shouldn't be blocked by that. Leader should be fine in most cases with being blocked, but we should analyze that more thoroughly.

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:107]: done

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:262]: done

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/testonly.rs:221]: done

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/testonly.rs:286]: done

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/tests.rs:42]: done

File: core/lib/zksync_core/src/consensus/testonly.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/testonly.rs:21]: done

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:340]: done

File: core/lib/zksync_core/src/consensus/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/mod.rs:33]: Nit: Given that we propagate errors here, should this use `anyhow::ensure!` instead?

File: core/lib/zksync_core/src/consensus/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/mod.rs:22]: Nit: Please derive `Debug` as we've discussed.

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (slowli) [inline core/lib/zksync_core/src/consensus/tests.rs:37]: Nit: Please move this constant next to `OPERATOR_ADDRESS`, as we've discussed.

File: core/lib/zksync_core/src/sync_layer/gossip/buffered/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/sync_layer/gossip/buffered/mod.rs:329]: IMO, it would be better to implement it as a separate trait (subtrait of `WriteBlockStore`, maybe) since this functionality indeed seems specific to validator nodes. OTOH, this could perform sanity checks and the hypothetical upstream caller in the consensus library would panic on an error (which would mean that the node cannot interpret the received block, or that the supermajority of validators has gone malicious).

File: core/lib/zksync_core/Cargo.toml
Comment (slowli) [inline core/lib/zksync_core/Cargo.toml:47]: Nit: Could it make sense to re-export the relevant items (i.e., `PayloadSource` for now) from the `executor` crate? From the usability perspective, it'd make sense to not require an additional direct dependency just for a single item.

**Review by slowli — APPROVED** [review]

File: core/lib/dal/src/blocks_dal.rs
Comment (pompon0) [reply core/lib/dal/src/blocks_dal.rs:415]: done

File: core/lib/zksync_core/src/consensus/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/mod.rs:22]: done

File: core/lib/zksync_core/src/consensus/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/mod.rs:33]: done

File: core/lib/zksync_core/src/consensus/tests.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/tests.rs:37]: done

File: core/lib/zksync_core/src/sync_layer/gossip/buffered/mod.rs
Comment (pompon0) [reply core/lib/zksync_core/src/sync_layer/gossip/buffered/mod.rs:329]: ack

File: core/lib/zksync_core/src/sync_layer/tests.rs
Comment (pompon0) [reply core/lib/zksync_core/src/sync_layer/tests.rs:2]: reverted

File: core/lib/zksync_core/Cargo.toml
Comment (pompon0) [reply core/lib/zksync_core/Cargo.toml:47]: ack

File: core/lib/zksync_core/src/consensus/storage.rs
Comment (pompon0) [reply core/lib/zksync_core/src/consensus/storage.rs:276]: done

**Review by brunoffranca — DISMISSED** [review]

**Review by brunoffranca — APPROVED** [review]

**Review by brunoffranca — APPROVED** [review]

**Pull request merged**
